##  Import libraries 

In [1]:
# Code to import libraries as you need in this assessment
import pandas as pd
import numpy as np
import re
import random


### 1. Examining and loading data
Examine "\<student\_id\>_dataset2.csv" the structure and schema. 

In [2]:
# Code to inspect the provided data file
# Couple of notes for all code block in this notebook
## please provide proper comment on your code
## Please re-start and run all cells to make sure codes are runable and include your output in the submission
df3 = pd.read_csv('s3972733_dataset2.csv')
df12 = pd.read_csv('s3972733_dataset1_solution.csv')
print(df3.shape)
print(df12.shape)

(5000, 9)
(107, 11)


### 2. Resolving schema conflicts

In [3]:
df12.head()

Id                                            Title   Location  \
0  70761086                     Antipodean Teachers Urgently     London   
1  71095030                 Deputy Hospital Manager  Bedford    Bedford   
2  72221399                                 Licensee Auditor  Berkshire   
3  51285926                     Home Manager Sheffield ****K  Sheffield   
4  69173865  Air Conditioning & Refrigeration (North London)         UK   

                                 Company   ContractType   ContractTime  \
0                       Sugarman Medical  non-specified      full_time   
1                         Brookdale Care      full_time  non-specified   
2                                   Hays      full_time      full_time   
3                          non-specified  non-specified  non-specified   
4  Riverside Technical Solutions Limited  non-specified      full_time   

                    Category   Salary             OpenDate  \
0              Teaching Jobs  36000.0  2013-07-13 00:00:00   
1  Healthcare & Nursing Jobs  60000.0  2012-07-21 00:00:00   
2  Accounting & Finance Jobs  42500.0  2012-01-18 15:00:00   
3  Healthcare & Nursing Jobs  37500.0  2012-04-27 00:00:00   
4           Engineering Jobs  32000.0  2013-05-08 12:00:00   

             CloseDate        SourceName  
0  2013-08-12 00:00:00  cv-library.co.uk  
1  2012-09-19 00:00:00    staffnurse.com  
2  2012-03-18 15:00:00           GAAPweb  
3  2012-05-11 00:00:00    careworx.co.uk  
4  2013-08-06 12:00:00     totaljobs.com

# Conflicts
- Monthly Payment -> Salary X
- Closing -> CloseDate X
- Organistation -> Company X
- Full-Time Equivalent -> Contract Time? X
- Opening -> OpenDate X
- Type -> Contract Type X
- Type - Values - NaN, etc: check values X
- Create New Id Numbers - randomnum - check if its in df12.Id, if not assign it X

In [4]:
# Code to inspect data and identify conflicts in schema
df3.head()

Monthly Payment              Closing               Organisation  \
0          1200.00  2012-12-27 12:00:00         Protocol Education   
1          3125.00  2013-08-25 15:00:00                Cordius Ltd   
2          4333.33  2013-01-16 15:00:00  Understanding Recruitment   
3          3750.00  2013-02-13 15:00:00     Wallace Hind Selection   
4          2916.67  2013-01-12 12:00:00        Greenhill Group Ltd   

           Location                Category  Full-Time Equivalent (FTE)  \
0                UK               Education                         0.8   
1                UK  Information Technology                         1.0   
2  Walton-On-Thames  Information Technology                         1.0   
3                UK                   Sales                         1.0   
4   Eastern England                   Sales                         1.0   

               Opening                                          Job Title  \
0  2012-11-27 12:00:00             Higher Level Teaching Assistant (HLTA)   
1  2013-06-26 15:00:00  Software Developers / Software Architects  Not...   
2  2012-10-18 15:00:00                      Senior BI Consultant (Cognos)   
3  2012-12-15 15:00:00  Technical Sales Manager, Pigments Fillers to S...   
4  2012-12-29 12:00:00                  Channel Account Manager  Software   

        Type  
0        NaN  
1  Permanent  
2  Permanent  
3        NaN  
4        NaN

### Conflict 1
- Renaming Columns

In [5]:
# Renaming Columns
df3.columns = ['Salary', 'CloseDate','Company','Location','Category', 'ContractTime', 'OpenDate', 'Title', 'ContractType']
df3.head()

Salary            CloseDate                    Company          Location  \
0  1200.00  2012-12-27 12:00:00         Protocol Education                UK   
1  3125.00  2013-08-25 15:00:00                Cordius Ltd                UK   
2  4333.33  2013-01-16 15:00:00  Understanding Recruitment  Walton-On-Thames   
3  3750.00  2013-02-13 15:00:00     Wallace Hind Selection                UK   
4  2916.67  2013-01-12 12:00:00        Greenhill Group Ltd   Eastern England   

                 Category  ContractTime             OpenDate  \
0               Education           0.8  2012-11-27 12:00:00   
1  Information Technology           1.0  2013-06-26 15:00:00   
2  Information Technology           1.0  2012-10-18 15:00:00   
3                   Sales           1.0  2012-12-15 15:00:00   
4                   Sales           1.0  2012-12-29 12:00:00   

                                               Title ContractType  
0             Higher Level Teaching Assistant (HLTA)          NaN  
1  Software Developers / Software Architects  Not...    Permanent  
2                      Senior BI Consultant (Cognos)    Permanent  
3  Technical Sales Manager, Pigments Fillers to S...          NaN  
4                  Channel Account Manager  Software          NaN

In [6]:
# Reference
df12.head()

Id                                            Title   Location  \
0  70761086                     Antipodean Teachers Urgently     London   
1  71095030                 Deputy Hospital Manager  Bedford    Bedford   
2  72221399                                 Licensee Auditor  Berkshire   
3  51285926                     Home Manager Sheffield ****K  Sheffield   
4  69173865  Air Conditioning & Refrigeration (North London)         UK   

                                 Company   ContractType   ContractTime  \
0                       Sugarman Medical  non-specified      full_time   
1                         Brookdale Care      full_time  non-specified   
2                                   Hays      full_time      full_time   
3                          non-specified  non-specified  non-specified   
4  Riverside Technical Solutions Limited  non-specified      full_time   

                    Category   Salary             OpenDate  \
0              Teaching Jobs  36000.0  2013-07-13 00:00:00   
1  Healthcare & Nursing Jobs  60000.0  2012-07-21 00:00:00   
2  Accounting & Finance Jobs  42500.0  2012-01-18 15:00:00   
3  Healthcare & Nursing Jobs  37500.0  2012-04-27 00:00:00   
4           Engineering Jobs  32000.0  2013-05-08 12:00:00   

             CloseDate        SourceName  
0  2013-08-12 00:00:00  cv-library.co.uk  
1  2012-09-19 00:00:00    staffnurse.com  
2  2012-03-18 15:00:00           GAAPweb  
3  2012-05-11 00:00:00    careworx.co.uk  
4  2013-08-06 12:00:00     totaljobs.com

### Conflict 2
- Converting Contract Type Values
    * nan -> non-specified
    * Permanent -> full_time
    * Fixed Term Contract -> part_time

In [7]:
# Exploring ContractType values
print(df3.ContractType.unique())
print(df12.ContractType.unique())

[nan 'Permanent' 'Fixed Term Contract']
['non-specified' 'full_time' 'part_time']


In [8]:
# Convert ContractType values
replace_ContractType = {np.nan:"non-specified", "Permanent":"full_time", "Fixed Term Contract":"part_time"}
df3.ContractType.replace(replace_ContractType, inplace=True)
print(df3.ContractType.unique())
df3.head()

['non-specified' 'full_time' 'part_time']


Salary            CloseDate                    Company          Location  \
0  1200.00  2012-12-27 12:00:00         Protocol Education                UK   
1  3125.00  2013-08-25 15:00:00                Cordius Ltd                UK   
2  4333.33  2013-01-16 15:00:00  Understanding Recruitment  Walton-On-Thames   
3  3750.00  2013-02-13 15:00:00     Wallace Hind Selection                UK   
4  2916.67  2013-01-12 12:00:00        Greenhill Group Ltd   Eastern England   

                 Category  ContractTime             OpenDate  \
0               Education           0.8  2012-11-27 12:00:00   
1  Information Technology           1.0  2013-06-26 15:00:00   
2  Information Technology           1.0  2012-10-18 15:00:00   
3                   Sales           1.0  2012-12-15 15:00:00   
4                   Sales           1.0  2012-12-29 12:00:00   

                                               Title   ContractType  
0             Higher Level Teaching Assistant (HLTA)  non-specified  
1  Software Developers / Software Architects  Not...      full_time  
2                      Senior BI Consultant (Cognos)      full_time  
3  Technical Sales Manager, Pigments Fillers to S...  non-specified  
4                  Channel Account Manager  Software  non-specified

### Conflict 3
- Converting ContractTime values
- **Assumptions**
    * non-specified = 0
    * full_time = 1
    * part_time = 0.1 <-> 0.9

In [9]:
# Convert ContractTime values
print(df3.ContractTime.value_counts())

# I'm assuming 0 = non-specified, 1 = full_time, anything in between = part_time
replace_ContractTime = {1.0:"full_time", 0.4:"part_time",0.8:"part_time", 0.6:"part_time", 0.2:"part_time"}
df3.ContractTime.replace(replace_ContractTime, inplace=True)
print(df3.ContractTime.value_counts())


ContractTime
1.0    4834
0.4      47
0.8      41
0.6      39
0.2      39
Name: count, dtype: int64
ContractTime
full_time    4834
part_time     166
Name: count, dtype: int64


Tell me about how you fix the conflict here

### Conflict 4
- Dataset 2 (df3) Needs an Id
- Randomly Generate 8 digit numbers
- Check to make sure it doesn't exist in Dataset 1 (df12)

In [10]:
# creating unique Id numbers with the same length digits
df3['Id'] = [random.randint(10_000_000, 99_999_999) for x in range(len(df3))]
df12id = set(df12.Id)
df3id = []
for x in range(len(df3)):
    while True:
        newId = random.randint(10_000_000, 99_999_999)
        if newId not in df12id:
            df3id.append(newId)
            df12id.add(newId)
            break        
df3 = df3[['Id','Title', 'Location', 'Company', 'ContractType','ContractTime','Category', 'Salary', 'OpenDate', 'CloseDate']]

df3.Id.unique()


array([31532062, 87732704, 76228472, ..., 96083177, 24700248, 94351129])

In [11]:
df3.head()

Id                                              Title  \
0  31532062             Higher Level Teaching Assistant (HLTA)   
1  87732704  Software Developers / Software Architects  Not...   
2  76228472                      Senior BI Consultant (Cognos)   
3  76769196  Technical Sales Manager, Pigments Fillers to S...   
4  28270559                  Channel Account Manager  Software   

           Location                    Company   ContractType ContractTime  \
0                UK         Protocol Education  non-specified    part_time   
1                UK                Cordius Ltd      full_time    full_time   
2  Walton-On-Thames  Understanding Recruitment      full_time    full_time   
3                UK     Wallace Hind Selection  non-specified    full_time   
4   Eastern England        Greenhill Group Ltd  non-specified    full_time   

                 Category   Salary             OpenDate            CloseDate  
0               Education  1200.00  2012-11-27 12:00:00  2012-12-27 12:00:00  
1  Information Technology  3125.00  2013-06-26 15:00:00  2013-08-25 15:00:00  
2  Information Technology  4333.33  2012-10-18 15:00:00  2013-01-16 15:00:00  
3                   Sales  3750.00  2012-12-15 15:00:00  2013-02-13 15:00:00  
4                   Sales  2916.67  2012-12-29 12:00:00  2013-01-12 12:00:00

## 3. Merging data

In [12]:
# Code to merge two data sets
df = pd.concat([df12, df3], axis=0)
df.head()

Id                                            Title   Location  \
0  70761086                     Antipodean Teachers Urgently     London   
1  71095030                 Deputy Hospital Manager  Bedford    Bedford   
2  72221399                                 Licensee Auditor  Berkshire   
3  51285926                     Home Manager Sheffield ****K  Sheffield   
4  69173865  Air Conditioning & Refrigeration (North London)         UK   

                                 Company   ContractType   ContractTime  \
0                       Sugarman Medical  non-specified      full_time   
1                         Brookdale Care      full_time  non-specified   
2                                   Hays      full_time      full_time   
3                          non-specified  non-specified  non-specified   
4  Riverside Technical Solutions Limited  non-specified      full_time   

                    Category   Salary             OpenDate  \
0              Teaching Jobs  36000.0  2013-07-13 00:00:00   
1  Healthcare & Nursing Jobs  60000.0  2012-07-21 00:00:00   
2  Accounting & Finance Jobs  42500.0  2012-01-18 15:00:00   
3  Healthcare & Nursing Jobs  37500.0  2012-04-27 00:00:00   
4           Engineering Jobs  32000.0  2013-05-08 12:00:00   

             CloseDate        SourceName  
0  2013-08-12 00:00:00  cv-library.co.uk  
1  2012-09-19 00:00:00    staffnurse.com  
2  2012-03-18 15:00:00           GAAPweb  
3  2012-05-11 00:00:00    careworx.co.uk  
4  2013-08-06 12:00:00     totaljobs.com

### Checking Duplicates
    - Id
    - Title

In [13]:
print(df.Id.duplicated().sum())
print(df.Title.duplicated().sum())
print(df.duplicated(subset=['Title', 'Location']).sum())
print(df.Category.value_counts())
print()
# pd.set_option('display.max_rows', None)
pd.reset_option('display.max_rows')
print(df.Location.value_counts())

0
0
0
Category
Information Technology              1317
Health                               789
Engineering                          744
Finance                              668
Sales                                498
Hospitality                          399
Education                            326
Marketing                            259
Accounting & Finance Jobs             25
Engineering Jobs                      19
Healthcare & Nursing Jobs             18
IT Jobs                               14
Teaching Jobs                         10
PR, Advertising & Marketing Jobs       9
Sales Jobs                             7
Hospitality & Catering Jobs            5
Name: count, dtype: int64

Location
UK                    772
London                671
South East London     314
The City              125
Central London         75
                     ... 
St. Ives                1
Bognor Regis            1
Hove                    1
Sutton-In-Ashfield      1
Tewkesbury              1
Name: c

### Merge Conflict 1: Need to Convert Categories
- Finance --> Accounting & Finance Jobs           X
- Engineering --> Engineering Jobs                X   
- Health --> Healthcare & Nursing Jobs            X
- Information Technology --> IT Jobs              X               
- Education --> Teaching Jobs                     X  
- Marketing --> PR, Advertising & Marketing Jobs  X  
- Sales --> Sales Jobs                            X
- Hospitality --> Hospitality & Catering Jobs     X    

In [14]:
replace_categories = {'Finance':'Accounting & Finance Jobs','Engineering':'Engineering Jobs','Health':'Healthcare & Nursing Jobs', 'Information Technology':'IT Jobs', 'Education':'Teaching Jobs', 'Marketing':'PR, Advertising & Marketing Jobs','Sales':'Sales Jobs','Hospitality':'Hospitality & Catering Jobs'}
df.Category.replace(replace_categories, inplace=True)
df.Category.value_counts()

Category
IT Jobs                             1331
Healthcare & Nursing Jobs            807
Engineering Jobs                     763
Accounting & Finance Jobs            693
Sales Jobs                           505
Hospitality & Catering Jobs          404
Teaching Jobs                        336
PR, Advertising & Marketing Jobs     268
Name: count, dtype: int64

In [15]:
df.Salary.value_counts()

Salary
2500.00    192
2916.67    178
3333.33    164
3750.00    151
2708.33    129
          ... 
3959.17      1
2326.58      1
2186.42      1
2753.33      1
983.17       1
Name: count, Length: 748, dtype: int64

### Merge Conflict 2: Converting Data Types
- Id --> int32
- Location --> string
- Company  --> string
- ContractType --> string
- ContractTime --> string
- Category --> string
- Salary --> float
- SourceName --> string

In [16]:
df.Id = df.Id.astype('int32')
df.Title = df.Title.astype('string')
df.Location = df.Location.astype('string')
df.Company = df.Company.astype('string')
df.ContractType = df.ContractType.astype('string')
df.ContractTime = df.ContractTime.astype('string')
df.Category = df.Category.astype('string')
df.Salary = df.Salary.astype('float')
df.SourceName = df.SourceName.astype('string')
df.dtypes

Id                       int32
Title           string[python]
Location        string[python]
Company         string[python]
ContractType    string[python]
ContractTime    string[python]
Category        string[python]
Salary                 float64
OpenDate                object
CloseDate               object
SourceName      string[python]
dtype: object

### Merge Conflict 3: Fill Empty SourceName Values
- Only 107 - should be the same size as df.shape (5107)

In [17]:
print(df.SourceName.value_counts())
print(df.shape)

SourceName
cv-library.co.uk          1
planetrecruit.com         1
GAAPweb                   1
careworx.co.uk            1
totaljobs.com             1
                         ..
jobsinsocialwork.co.uk    1
cardandpaymentjobs.com    1
cvjobstore.com            1
workthing.com             1
thegraduate.co.uk         1
Name: count, Length: 107, dtype: Int64
(5107, 11)


In [18]:
replace_SourceName = {np.NaN:"non-specified"}
df.SourceName.replace(replace_SourceName, inplace=True)
print(df.SourceName.value_counts())
print(df.shape)

SourceName
non-specified            5000
jobs.guardian.co.uk         1
GAAPweb                     1
careworx.co.uk              1
totaljobs.com               1
                         ... 
cvjobstore.com              1
workthing.com               1
thegraduate.co.uk           1
emptylemon.co.uk            1
jobs.newstatesman.com       1
Name: count, Length: 108, dtype: Int64
(5107, 11)


In [19]:
df.head()

Id                                            Title   Location  \
0  70761086                     Antipodean Teachers Urgently     London   
1  71095030                 Deputy Hospital Manager  Bedford    Bedford   
2  72221399                                 Licensee Auditor  Berkshire   
3  51285926                     Home Manager Sheffield ****K  Sheffield   
4  69173865  Air Conditioning & Refrigeration (North London)         UK   

                                 Company   ContractType   ContractTime  \
0                       Sugarman Medical  non-specified      full_time   
1                         Brookdale Care      full_time  non-specified   
2                                   Hays      full_time      full_time   
3                          non-specified  non-specified  non-specified   
4  Riverside Technical Solutions Limited  non-specified      full_time   

                    Category   Salary             OpenDate  \
0              Teaching Jobs  36000.0  2013-07-13 00:00:00   
1  Healthcare & Nursing Jobs  60000.0  2012-07-21 00:00:00   
2  Accounting & Finance Jobs  42500.0  2012-01-18 15:00:00   
3  Healthcare & Nursing Jobs  37500.0  2012-04-27 00:00:00   
4           Engineering Jobs  32000.0  2013-05-08 12:00:00   

             CloseDate        SourceName  
0  2013-08-12 00:00:00  cv-library.co.uk  
1  2012-09-19 00:00:00    staffnurse.com  
2  2012-03-18 15:00:00           GAAPweb  
3  2012-05-11 00:00:00    careworx.co.uk  
4  2013-08-06 12:00:00     totaljobs.com

#### Finding global key for the data
Find a proper global key for the integrated data and give justification here.

When looking for a global key, we should make sure the key is:
 - Unique: Each record should have a unique key.
 - Not null: They key should never be null or empty.
 - Immutable: The value of the key should not change.
 - Consistent across all datasets: They key should be available and mean the same across all datasets.
 - Simplistic: Prefer a single column over a composite key if possible.

The Id column of the merged dataframe (df.Id) is the global key for the data. Each number is unique, not null, and should not be mutable. The 8 digits are also simplistic and in a single column. The Id can also Identify each of the title. Using 8 digits keys will also make the dataset future-proof. 


In [20]:
# Code to identify global key
global_key = df.Id

### 5. Saving the integrated and reshaped data
The last part of the integration process is to export our output data to csv format, named as:
- '\<student\_id\>_dataset_integrated.csv'

In [21]:
# code to save output data
df.to_csv('dataset_integrated.csv')

## Summary

I began by importing the libraries that I anticipated needing. Subsequently, I imported two CSV files: dataset1_solution.csv and dataset2.csv. Prior to diving into the coding aspect, I outlined a list of issues and potential conflicts within the data, using markdown blocks in a manner akin to a TODO list in note-taking. I adopted a top-down approach, initially focusing on aligning the column names between the two datasets. I then proceeded to examine each column individually to understand the format and content of the values. I noticed that the "ContractType" column had missing data and inconsistent terminology. To address this, I standardized the values and filled in the missing entries to align with dataset1. A similar process was required for the "ContractTime" column, albeit with different values to consider. Next, I generated an "Id" column for dataset2. Given that the "Id" would serve as my global key, it was important that it be an 8-digit, unique identifier that did not overlap with any Ids in dataset1. After assigning unique Ids to dataset2, I merged the two datasets. After merging, I focused on identifying and resolving duplicate entries. I found that the "Category" column contained redundant names, which I then standardized to match those in dataset1. I also observed that the locations in dataset2 were based in Australia and that the salaries were represented in floating-point format. I converted all columns to their appropriate data types. Lastly, I discovered that the "SourceName" column had missing entries. I filled these in to align with dataset1, confirmed the global key, and exported the unified dataframe into a new CSV file.